ports/mimxrt/machine_i2c.c

```diff
@@ -0,0 +1,217 @@
+        clk_init = false;
+        // Set clock source for LPI2C
+        CLOCK_SetMux(kCLOCK_Lpi2cMux, LPI2C_CLOCK_SOURCE_SELECT); // USB1 PLL (480 MHz)
+        CLOCK_SetDiv(kCLOCK_Lpi2cDiv, LPI2C_CLOCK_SOURCE_DIVIDER);
+    }
+
+    // Initialise the I2C peripheral if any arguments given, or it was not initialised previously.
+    lpi2c_set_iomux(self->i2c_hw_id, drive);
+    self->master_config = m_new_obj(lpi2c_master_config_t);
+    LPI2C_MasterGetDefaultConfig(self->master_config);
+    // Initialise the I2C peripheral.
+    self->master_config->baudRate_Hz = args[ARG_freq].u_int;
+    LPI2C_MasterInit(self->i2c_inst, self->master_config, LPI2C_CLOCK_FREQUENCY);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+static void lpi2c_master_callback(LPI2C_Type *base, lpi2c_master_handle_t *handle, status_t status, void *self_in) {
+    machine_i2c_obj_t *self = (machine_i2c_obj_t *)self_in;
+
+    self->transfer_busy = false;
+    self->transfer_status = status;
+}
+
+STATIC int machine_i2c_transfer_single(mp_obj_base_t *self_in, uint16_t addr, size_t len, uint8_t *buf, unsigned int flags) {
+    machine_i2c_obj_t *self = (machine_i2c_obj_t *)self_in;
+    status_t ret;
+    lpi2c_master_handle_t g_master_handle;
+    lpi2c_master_transfer_t masterXfer = {0};
+    LPI2C_MasterTransferCreateHandle(self->i2c_inst, &g_master_handle, lpi2c_master_callback, self);
+
+    if (flags & MP_MACHINE_I2C_FLAG_READ) {
+        masterXfer.direction = kLPI2C_Read;
+    } else {
+        masterXfer.direction = kLPI2C_Write;
+    }
+    if (addr < 0x80) {  // 7 or 10 bit address?
+        masterXfer.slaveAddress = addr;
+    } else {
+        masterXfer.slaveAddress = 0x78 | ((addr >> 8) & 0x03);
+        masterXfer.subaddress = addr & 0xff;
+        masterXfer.subaddressSize = 1;
+    }
+    masterXfer.data = buf;
+    masterXfer.dataSize = len;
+    if (flags & MP_MACHINE_I2C_FLAG_STOP) {
+        masterXfer.flags = kLPI2C_TransferDefaultFlag;
+    } else {
+        masterXfer.flags = kLPI2C_TransferNoStopFlag;
+    }
+    self->transfer_busy = true;
+
+    // Send master data to slave in non-blocking mode
+    ret = LPI2C_MasterTransferNonBlocking(self->i2c_inst, &g_master_handle, &masterXfer);
+    if (ret != kStatus_Success) {
+        return -MP_EIO;
+    }
+    //  Wait for the transfer to complete
+    while (self->transfer_busy) {
+        MICROPY_EVENT_POLL_HOOK
+    }
+
+    // Transfer will not send a stop in case of errors like NAK. So it#s done here.
+    if (flags & MP_MACHINE_I2C_FLAG_STOP && self->transfer_status != kStatus_Success) {
+        LPI2C_MasterStop(self->i2c_inst);
+    }
+
+    if (self->transfer_status == kStatus_Success) {
+        return len;
+    } else if (self->transfer_status == kStatus_LPI2C_Nak) {
+        return -MP_ENODEV;
+    } else {
+        return -MP_EIO;
+    }
+}
+
+STATIC const mp_machine_i2c_p_t machine_i2c_p = {
+    .transfer = mp_machine_i2c_transfer_adaptor,
+    .transfer_single = machine_i2c_transfer_single,
```

Comment:
I guess you needed to use the adaptor because there was no easy way to transfer multiple segments?  It's not a problem, just wondering.

Reply:
It was just a copy of method wihtout much thinking.

Reply:
But in fact the NXP API performs all the frame handling like handling start and stop conditions,  and expect a single message to be sent.